### ISSUE TITLE
Representation hiding for class instances

### ISSUE THREAD

**Issue by adamgundry** [ghc-proposals/ghc-proposals#483]
If you have an abstract datatype, instances of certain classes may expose details of the representation, and thereby provide a way around the abstraction boundary. The most obvious case of this is `Generic`. Thus it would be nice to have a way to control whether downstream modules see an instance.

Something like this exists already for some built-in constraints:
 * `Coercible` looks through newtypes only if the newtype constructor is in scope
 * `HasField` is solved only if the corresponding record field selector is in scope

We could special-case `Generic`, but I think it would be nicer to have a language feature that can be used for arbitrary classes. I'm imagining being able to mark particular instances as "private", with the semantics that the instance is available only if the data constructor(s) of the type are in scope. (An interesting question is whether the instance is hidden completely, or if GHC should refuse to use it and report an error.)

For example:
```hs
module Internal where
  data AbstractType = MkAbstractType { foo :: Int }
    deriving private Generic

module External (AbstractType) where
  import Internal
  -- instance Generic AbstractType is available in this module, because the constructor was imported

module Client where
  import External
  -- instance Generic AbstractType is not available here, because the constructor was not imported
```

Does this sound worthwhile to anyone else?

See https://github.com/kcsongor/generic-lens/issues/142 for the discussion that prompted this ticket.

**Comment by chreekat** [comment]
I think this sounds like an interesting idea, but I also think it has the potential for a frustrating epidemic of misuse: people adding it to instances "because it's there". Particularly because the same keyword `private` is more or less obligatory boilerplate in some languages.

That's just a minor comment though.

**Comment by adamgundry** [comment]
Yes, I did wonder if we should have `public` for the normal behaviour, `protected` for requiring the data constructor to be in scope and `private` for visible within the same module only. Or some such trichotomy. But that might take the resemblance to certain boilerplate-heavy languages too far. :laughing:
